pkg/binding/binding.go

```diff
@@ -109,6 +109,7 @@ func (c *Controller) syncBinding(ctx context.Context, bindingName string) error
 	policyWithStatus := policy.DeepCopy()
 	policyWithStatus.Status = v1alpha1.BindingPolicyStatus{
 		ObservedGeneration: policy.Generation,
+		Conditions:         policy.Status.Conditions,
```

Comment:
This is going backwards. Since Binding is derived from BindingPolicy, the flow should be: BindingPolicy desired state -> Binding desired state -> stuff happens -> Binding reported state -> BindingPolicy reported state.

Reply:
We already have reported state flowing from Binding to BindingPolicy in https://github.com/kubestellar/kubestellar/blob/v0.26.0-alpha.1/pkg/binding/binding.go#L91 through https://github.com/kubestellar/kubestellar/blob/v0.26.0-alpha.1/pkg/binding/binding.go#L112

Reply:
> Are you suggesting that, that logic in SyncBindingPolicy should be moved here into SyncBinding to fit into the flow you described?

No. My main point was that there is already a flow updating the BindingPolicy's status and we should not introduce a second one. Here a a few principles to keep in mind: (a) when updating status, update `.status.reportedGeneration` so that the reader knows what the controller was looking at when doing that update; (b) for a given atom in a given API object, only one controller should author it; and (c) in a given controller, only one goroutine at a time should be writing to a given API object (with "subresources" considered distinct objects for this purpose).

The current code structure is dissatisfying to me. A Binding object is derived from a BindingPolicy object, is essentially the implementation of a BindingPolicy object. Thus, the logical flow is BindingPolicy desired state -> Binding desired state -> stuff happens --> Binding reported state -> BindingPolicy reported state. However, the history of the development of this code leaves us in a different place. Originally BindingPolicy was implemented directly; later Binding objects were introduced as a side-effect of BindingPolicy. I think this is an inferior view, I prefer the one I outlined at the start of this paragraph. But the code structure is not fully there yet. But I think that we should be moving in that direction.

The latest draft of this PR introduces logic in the binding controller that is redundant with logic in the status controller. The status controller already reacts to informer events about StatusCollector objects (in `func (c *combinedStatusResolver) NoteStatusCollector`) by going through all the Binding objects and reacting to the current state of that StatusCollector. I think that the simplest and most consistent thing would be for this processing to be extended so that the condition gets reported first in the Binding's Conditions and then this, along with the Errors, propagates to the BindingPolicy's Status.

It seems to me that the current split between "status controller" and "binding controller" is spurious. Both are concerned with implementing aspects of BindingPolicy and Binding objects. I think the split was originally motivated by the observation that the status functionality can just react to informer events about WorkStatus objects. I think that is not true; we see that fully implementing all the proper reactivity involves a tight integration between the two controllers, casting doubt on whether they should be two separate controllers.

Reply:
Thanks for sharing these insights and thoughts!

On the 'redundant' logic: I tried to squeeze the logic into `func (c *combinedStatusResolver) NoteStatusCollector`, as shown in experimental commit https://github.com/kubestellar/kubestellar/pull/2643/commits/52b2b0aea4da2450367c9a1f461ed7567e24229a. The resulting code looks a little awkward to me because of three reasons. (a) the signature of the method has to be changed a lot to make it work, making it sort of overstuffed. The method should be a highly cohesive and internal thing. (b) there is no logging and error handling inside the method. (c) Most importantly, the experimental commit doesn't work because of #2687 .

On the 'redundant' logic, too: I don't think the logic in binding controller is redundant, because the informer events from _both_ bindingpolicies and statuscollectors should trigger the updates of the Condition. For example, say BindingPolicy `bp` requires StatusCollector `sc`, and user creates `bp` but hasn't created `sc` yet. At this time the user should be able to see the Condition of `bp`, without any informer events from statuscollectors.

On the split of the binding controller and the status controller: I fully agree that the split is obsolete now. During my coding I felt more and more data and logic are shared tightly between the two. One example is the shared 'broker' from binding controller to status controller, second example is the shared listers for workload objects, third but not last example is the logic for updating the Condition in this PR.

Reply:
BTW, I have a question regarding (b) in the listed principles.
> (b) for a given atom in a given API object, only one controller should author it

I remember that k8s allows multiple field managers to co-own a field. Is this feature conflicting with principle (b)?

Reply:
> k8s allows multiple field managers to co-own a field

Yes, but in my opinion only one should be the decider of what the value is. Another might be in the habit of writing the same value, but even this gets tricky to keep out of trouble. In my opinion each atom [_should_](https://datatracker.ietf.org/doc/html/rfc2119) have one author.

Reply:
I looked at commit 52b2b0aea4da2450367c9a1f461ed7567e24229a, and it puts more in NoteStatusCollector than I was thinking. I think that an approach that is more consistent with the overall style of these controllers would be to expand what is in a `combinedStatusResolution` to include a set of names of missing StatusCollector objects. The NoteStatusCollector method would update these sets incrementally, and the existing path for writing BindingStatus would naturally be extended with this information.

Reply:
> the experimental commit doesn't work because of #2687 .

So I opened #2690 to look into that. The first commit changed the testing, and that exposed a bug. Later commits investigate and fix the bug. The fix does the kernel of what I described above, making `combinedStatusResolution` hold information about all relevant StatusCollector objects --- including ones that do not exist at the moment.